// devicer.js/dist/main.js
export { getHash, compareHashes } from "./libs/tlsh.js";
export { calculateConfidence, createConfidenceCalculator, calculateScoreBreakdown, computeAdaptiveStabilityWeights, computeAttractorRisk, computeEntropyContribution, computeEvidenceRichness, computeFieldAgreement, computeMissingBothSides, computeMissingOneSide, computeStructuralStability } from "./libs/confidence.js";
export { registerComparator, registerWeight, registerPlugin, unregisterComparator, unregisterWeight, setDefaultWeight, clearRegistry, initializeDefaultRegistry } from "./libs/registry.js";
export { createInMemoryAdapter } from "./libs/adapters/inmemory.js";
export { createSqliteAdapter } from "./libs/adapters/sqlite.js";
export { createPostgresAdapter } from "./libs/adapters/postgres.js";
export { createRedisAdapter } from "./libs/adapters/redis.js";
export { DeviceManager } from "./core/DeviceManager.js";
export { PluginRegistrar } from "./core/PluginRegistrar.js";
export { AdapterFactory } from "./core/AdapterFactory.js";
export { defaultLogger, defaultMetrics } from "./libs/default-observability.js";
